misc fixes to `flatten`

Krastanov opened the pull request:
Fixes https://github.com/JuliaLang/julia/issues/50710
Fixes #196
Fixes #194
Closes #195

cstjean commented on src/utils.jl:
```diff
@@ -458,16 +458,46 @@ function flatten1(ex)
+flatten(x) = x
+function flatten(x::Expr)
+  if isexpr(x, :try) # try args can be either false or blocks, but not non-block exprs, so we need to be a bit verbose here
+    isa(x.args[2], Symbol) || x.args[2] == false || error("MacroTools.flatten has encountered a misformed `try` block. Please report this as a bug.")
```
```suggestion
    isa(x.args[2], Symbol) || x.args[2] == false || error("Misformed `try` block.")
```

Krastanov replied:
I would prefer to leave this verbose. I have strong negative feelings to brief error messages. Error messages should be detailed, and should tell you what you can do to fix the error. For what is worth, I believe the SciML and Rust communities are a good example of this attitude.

cstjean commented on src/utils.jl:
```diff
@@ -458,16 +458,46 @@ function flatten1(ex)
   return length(ex′.args) == 1 ? ex′.args[1] : ex′
 end
 
+false_or_bflatten(x) = x==false ? false : bflatten(x)
+
+function bflatten(x)
+    fx = flatten(x)
+    return isexpr(fx, :block) ? fx : Expr(:block,fx)
+end
```
```suggestion
# Helper - only flatten the arguments of x
flatten_args(x::Expr) = Expr(x.head, map(flatten, x.args)...)
flatten_args(x) = x
```
I think you could replace all `bflatten` or `false_or_bflatten` calls with this? Like I mentioned in the other threads, I don't want to special-case `false`, I think that's mostly a Julia mistake. If we simply preserve `block`s inside of trys, it should be fine.

Krastanov replied:
The arguments of try have to be blocks, even if they contain only one argument. This would break that convention (and potentially break pretty printers and other macro code that relies on the convention). I can not seem to be able to force julia to parse a try expression into something that does not have the arguments either as blocks or as false. I think this is done on purpose, to disambiguate `finally false end` from complete absence of a finally.

```
julia> :(try; catch; else; finally; end) |> striplines |> dump
Expr
  head: Symbol try
  args: Array{Any}((5,))
    1: Expr
      head: Symbol block
      args: Array{Any}((0,))
    2: Bool false
    3: Expr
      head: Symbol block
      args: Array{Any}((0,))
    4: Expr
      head: Symbol block
      args: Array{Any}((0,))
    5: Expr
      head: Symbol block
      args: Array{Any}((0,))

julia> :(try; g() catch; h() else; f() finally; m() end) |> striplines |> dump
Expr
  head: Symbol try
  args: Array{Any}((5,))
    1: Expr
      head: Symbol block
      args: Array{Any}((1,))
        1: Expr
          head: Symbol call
          args: Array{Any}((1,))
            1: Symbol g
    2: Bool false
    3: Expr
      head: Symbol block
      args: Array{Any}((1,))
        1: Expr
          head: Symbol call
          args: Array{Any}((1,))
            1: Symbol h
    4: Expr
      head: Symbol block
      args: Array{Any}((1,))
        1: Expr
          head: Symbol call
          args: Array{Any}((1,))
            1: Symbol m
    5: Expr
      head: Symbol block
      args: Array{Any}((1,))
        1: Expr
          head: Symbol call
          args: Array{Any}((1,))
            1: Symbol f
```

cstjean replied:
> The arguments of try have to be blocks, even if they contain only one argument. This would break that convention (and potentially break pretty printers and other macro code that relies on the convention). I can not seem to be able to force julia to parse a try expression into something that does not have the arguments either as blocks or as false. I think this is done on purpose, to disambiguate finally false end from complete absence of a finally.

Doesn't my suggested code preserve blocks? `Expr(x.head, map(flatten, x.args)...)`? Maybe I'm misunderstanding.

cstjean commented on src/utils.jl:
```diff
@@ -458,16 +458,46 @@ function flatten1(ex)
+function flatten(x::Expr)
+  if isexpr(x, :try) # try args can be either false or blocks, but not non-block exprs, so we need to be a bit verbose here
+    isa(x.args[2], Symbol) || x.args[2] == false || error("MacroTools.flatten has encountered a misformed `try` block. Please report this as a bug.")
+    if length(x.args) == 3
+      # try catch end # 3 args, 2nd is a symbol or false, 3rd is block
+      return Expr(x.head, bflatten(x.args[1]), x.args[2], bflatten(x.args[3]))
+    elseif length(x.args) == 4
+      # try finally end # 4 args, 2nd==3rd==false, 4th is block
+      # try catch finally end # 4 args, 2nd is a symbol or false, 3rd and 4th are block
+      return Expr(x.head, bflatten(x.args[1]), x.args[2], false_or_bflatten(x.args[3]), bflatten(x.args[4]))
+    elseif length(x.args) == 5
+      # try catch else end # 5 args, 2nd is a symbol or false, 3rd is block, 4th is false, 5th is block
+      # try catch else finally end # 5 args, 2nd is a symbol or false, the rest are block
+      return Expr(x.head, bflatten(x.args[1]), x.args[2], bflatten(x.args[3]), false_or_bflatten(x.args[4]), bflatten(x.args[5]))
+    else
+      error("MacroTools.flatten has encountered a misformed `try` block. Please report this as a bug.")
+    end
```
```suggestion
    else
        return Expr(x.head, map(flatten_args, x.args)...)
    end
```
I think this is very appealing in simplicity --- if it works! It's also easy to explain, `flatten flattens blocks, except those in try clauses, because they have a special meaning`. Yes, technically we can flatten some more in some cases, but I don't think that there's any code out there which relies on this fact, but which isn't also susceptible to the bugs we've been discussing.

Krastanov replied:
I think the requirement for try arguments to be either blocks or false (discussed in the comment above) precludes this nicer more elegant treatment.

cstjean replied:
`flatten_args(false)` is `false`, and `flatten_args(some_block)` is a block. Why wouldn't it work?

Krastanov replied:
You are right, that would work, but I think this would not be flattened:

```
try
    begin
        f()
        g()
    end
catch
end
```

while at least ResumableFunctions expects this to be flattened.

I implemented what you suggested in the last few minutes and that is how I found the issue.

cstjean replied:
True. Maybe `return Expr(:try, flatten(x.args[1]), map(flatten_args, x.args[2:end])...)` ? But then, that is also problematic in the other arguments. 🤔 

Would be good to add an extra test case.

Maybe we need to write `flatten_but_keep_block` instead of `flatten_args`?  Something like `flatten_but_keep_block (x) = (y = flatten(x); isexpr(x, :block) && !isexpr(y, :block) ? Expr(:block, y) : y)`. Or maybe there's a less wasteful implementation.

Krastanov replied:
yeah, I am leaning in the same direction on all of your suggestions. I will plan to send you something over the weekend.

cstjean commented on test/flatten_try.jl:
```diff
@@ -0,0 +1,30 @@
+        quote try; f(); catch; else; end; end,
+        quote try; f(); catch; 3+3; else; 2+2; end; end,
+        quote try; f(); catch E; else; end; end,
+        quote try; f(); catch E; 3+3; else; 2+2; end; end
```
Nice tests!

cstjean commented on test/runtests.jl:
```diff
@@ -8,4 +8,8 @@ include("split.jl")
 
+if VERSION>=v"1.8"
+    include("flatten_try.jl")
+end
```
I don't think we need to / should special-case here. The PR will work on older versions of Julia, and it's nice to fix #196 there too.

Krastanov replied:
The `try catch else` is not valid syntax for earlier versions of Julia. That is why I have some tests copied in utils.

Krastanov replied:
The tests in utils will execute on earlier versions of julia and take care of the examples that do not contain `else`

cstjean replied:
👍 Maybe just a `# because of new try/else syntax` comment then please.

cstjean reviewed (commented):
👍 Thank you for working on this!

Krastanov commented:
Thanks for the quick review, Cédric!

My comment for the error message is a personal preference, but of course I will change it if you confirm that you as a maintainer prefer the shorter style.

I think I have given a complete argument for why simplifying the treatment of the try block would not work. Let me know if I am missing something or if you want me to approach it differently.

There are newly added tests that run on all julia versions and more tests for >=1.8 which are the only versions with support for `try else`.

cstjean commented:
> My comment for the error message is a personal preference, but of course I will change it if you confirm that you as a maintainer prefer the shorter style.

I don't have a strong personal preference, but conciseness is a pretty big theme of [this package](https://github.com/search?q=repo%3AFluxML%2FMacroTools.jl%20error&type=code) and I'm inclined to follow it, if you don't mind.

In any case, I would at least take out the `Please report it`, because 1. it's implicit in all error messages and 2. a malformed try block isn't a priori MacroTools' fault.

Krastanov commented:
I will try to do a couple more changes, currently I reintroduced an issue with ResumableFunctions. Moving this to a draft.

Krastanov commented:
I think now all comments are addressed. The difficulty was that arg[2] had to be a symbol or false, unlike the arg[3] or arg[4] which have to be a block or false. This is what makes it difficult to just use a `map(some_kind_of_flatten, args)`.

Now it is as simple as I can think to make it.

Added clarification comments in code and tests. Added more tests. Fixed the error messages.

Please excuse the confused back and forth, I do not have much experience working with Julia macros.

cstjean commented on src/utils.jl:
```diff
@@ -462,9 +462,12 @@ function flatten1(ex)
+false_or_symbol_or_bflatten(x) = x==false || isa(x, Symbol) ? x : bflatten(x)
+function bflatten(x) # flatten down to a block (i.e. a single symbol is turned into `begin symbol end`, unlike `flatten` which would just return the symbol)
+    fx = flatten(x)
+    return isexpr(fx, :block) ? fx : Expr(:block,fx)
```
If I `flatten(Expr(:try, :(sin(10)), false))`, then it is pretty odd that `flatten` will _add_ a new `:block`, which wasn't there before. That's the opposite of what you want from `flatten`. If this were a brand new function, I'd say that it's OK because that's what the Julia parser yields anyway, but given that we'll potentially break some code, I'd rather try to stick to the previous behaviour here:

```suggestion
    return isexpr(fx, :block) || !isexpr(x, :block) ? fx : Expr(:block,fx)
```

cstjean replied:
Also, with this change, I don't think you need `false_or_symbol_or_bflatten`.

Krastanov replied:
I guess I did not want to be permissive, I wanted this to work only for `false` and `::Symbol`, because that is the only thing that Julia produces. But maybe someone would be making weird expressions, so I should not be that restrictive, as you mentioned above.

cstjean commented on src/utils.jl:
```diff
@@ -462,9 +462,12 @@ function flatten1(ex)
-flatten_args(x) = x
+# Helpers for flattening try blocks
+false_or_symbol_or_bflatten(x) = x==false || isa(x, Symbol) ? x : bflatten(x)
+function bflatten(x) # flatten down to a block (i.e. a single symbol is turned into `begin symbol end`, unlike `flatten` which would just return the symbol)
```
```suggestion
bflatten(x) = x
function bflatten(x::Expr) # flatten down to a block (i.e. a single symbol is turned into `begin symbol end`, unlike `flatten` which would just return the symbol)
```
then you don't need `false_or_symbol_or_bflatten`.

cstjean commented on src/utils.jl:
```diff
@@ -458,16 +458,38 @@ function flatten1(ex)
+
+flatten(x) = x
+function flatten(x::Expr)
+  if isexpr(x, :try) # try args can be either false or blocks, but not non-block exprs, except for the second argument that has to be false or symbol... so we need to be a bit verbose here
```
That comment is only true for the parser. Macro code can form expressions like `eval(Expr(:try, 1, false, 3))`, and julia takes it just fine. I don't think we should trigger a new error on an `Expr` like that.

Krastanov replied:
I will clarify the comment. I do not think we raise an error on your example, but we do modify it. I will add tests.

Krastanov commented:
Made the changes such that unnatural block-less try expressions are not forced to have a block inserted around them anymore and simplified the code as you suggested. Added tests for the block-less try expressions.

cstjean reviewed (approved):
💯 It's a really nice PR, thank you for bearing with me on the iterations!

Krastanov commented on src/utils.jl:
```diff
@@ -458,16 +458,38 @@ function flatten1(ex)
+
+flatten(x) = x
+function flatten(x::Expr)
+  if isexpr(x, :try) # for try blocks generated by the Julia parser, args can be either false or blocks, but not non-block exprs, except for the second argument that has to be false or symbol... so we need to be a bit verbose here
```
I concur

cstjean commented:
Should we consider this a breaking change? I think no, I think it's a bugfix, but still, could you please summarize the expressions whose outcome has changed? If you run the test suite on 1.9, what fails? 

Last I checked MacroTools is used in thousands of packages, so personally it's always a bit unnerving to tag a new release. I think we should sleep on this PR for a week or so in case we think of something wrong.

Don't hesitate to ping on this PR in case I forget about it. Typhoon's about to hit where I live...

Krastanov commented:
I think it should be tagged as a bug fix. For the following two bugs.

1. `flatten` was a bit fragile with `try` blocks which got parsed slightly differently on 1.10 (extra line number nodes), which triggered wrong behavior. I will make one more commit in which I document that a bit better with more comments.

2. I think it is also correct to say that `flatten(quote try f() catch end end) == Expr(:try, Expr(:call, ;f), false)` was a bug that is now fixed.

I suggest tagging as bug fix and only if there is a complaint we can yank the release from the registry and tag as breaking. I do not expect complaints though.

I agree about waiting a week.

Be safe in the storm!

Krastanov commented:
The last commit documents what issue existed both in 1.9 and 1.10 and what is the issue that existed only in 1.10.

All tests in CI pass.

All my local tests on 1.9 and nightly pass for MacroTools, ResumableFunctions, and ConcurrentSim (the original library in which I detected these issues).

The PR was reviewed and approved.

Should be good for merge and release as non-breaking bug fix after a week of due-diligence wait time.

Krastanov commented:
@cstjean , I hope the storm was not too bad! Bumping this in case it can be merged now.
